File: open-nucleus-app/lib/core/constants/api_paths.dart
/// Every REST endpoint path exposed by the Open Nucleus API gateway.
///
/// Static strings are used for fixed paths; static methods with parameters
/// are used for paths that contain resource IDs.  All paths start with `/`.
class ApiPaths {
  ApiPaths._(); // non-instantiable

  // ---------------------------------------------------------------------------
  // Health
  // ---------------------------------------------------------------------------

  static const String health = '/health';

  // ---------------------------------------------------------------------------
  // Auth
  // ---------------------------------------------------------------------------

  static const String authLogin = '/api/v1/auth/login';
  static const String authRefresh = '/api/v1/auth/refresh';
  static const String authLogout = '/api/v1/auth/logout';
  static const String authWhoami = '/api/v1/auth/whoami';

  // ---------------------------------------------------------------------------
  // Patients
  // ---------------------------------------------------------------------------

  static const String patients = '/api/v1/patients';
  static const String patientsSearch = '/api/v1/patients/search';
  static const String patientsMatch = '/api/v1/patients/match';

  static String patient(String id) => '/api/v1/patients/$id';
  static String patientHistory(String id) => '/api/v1/patients/$id/history';
  static String patientTimeline(String id) => '/api/v1/patients/$id/timeline';
  static String patientErase(String id) => '/api/v1/patients/$id/erase';

  // ---------------------------------------------------------------------------
  // Encounters
  // ---------------------------------------------------------------------------

  static String encounters(String patientId) =>
      '/api/v1/patients/$patientId/encounters';

  static String encounter(String patientId, String encounterId) =>
      '/api/v1/patients/$patientId/encounters/$encounterId';

  // ---------------------------------------------------------------------------
  // Observations
  // ---------------------------------------------------------------------------

  static String observations(String patientId) =>
      '/api/v1/patients/$patientId/observations';

  static String observation(String patientId, String observationId) =>
      '/api/v1/patients/$patientId/observations/$observationId';

  // ---------------------------------------------------------------------------
  // Conditions
  // ---------------------------------------------------------------------------

  static String conditions(String patientId) =>
      '/api/v1/patients/$patientId/conditions';

  static String condition(String patientId, String conditionId) =>
      '/api/v1/patients/$patientId/conditions/$conditionId';

  // ---------------------------------------------------------------------------
  // Medication Requests
  // ---------------------------------------------------------------------------

  static String medicationRequests(String patientId) =>
      '/api/v1/patients/$patientId/medication-requests';

  static String medicationRequest(String patientId, String requestId) =>
      '/api/v1/patients/$patientId/medication-requests/$requestId';

  // ---------------------------------------------------------------------------
  // Allergy Intolerances
  // ---------------------------------------------------------------------------

  static String allergyIntolerances(String patientId) =>
      '/api/v1/patients/$patientId/allergy-intolerances';

  static String allergyIntolerance(String patientId, String allergyId) =>
      '/api/v1/patients/$patientId/allergy-intolerances/$allergyId';

  // ---------------------------------------------------------------------------
  // Immunizations
  // ---------------------------------------------------------------------------

  static String immunizations(String patientId) =>
      '/api/v1/patients/$patientId/immunizations';

  static String immunization(String patientId, String immunizationId) =>
      '/api/v1/patients/$patientId/immunizations/$immunizationId';

  // ---------------------------------------------------------------------------
  // Procedures
  // ---------------------------------------------------------------------------

  static String procedures(String patientId) =>
      '/api/v1/patients/$patientId/procedures';

  static String procedure(String patientId, String procedureId) =>
      '/api/v1/patients/$patientId/procedures/$procedureId';

  // ---------------------------------------------------------------------------
  // Consents (patient-scoped)
  // ---------------------------------------------------------------------------

  static String patientConsents(String patientId) =>
      '/api/v1/patients/$patientId/consents';

  // ---------------------------------------------------------------------------
  // Consents (top-level)
  // ---------------------------------------------------------------------------

  static String consent(String consentId) => '/api/v1/consents/$consentId';

  static String consentVc(String consentId) =>
      '/api/v1/consents/$consentId/vc';

  // ---------------------------------------------------------------------------
  // Practitioners
  // ---------------------------------------------------------------------------

  static const String practitioners = '/api/v1/practitioners';

  static String practitioner(String id) => '/api/v1/practitioners/$id';

  // ---------------------------------------------------------------------------
  // Organizations
  // ---------------------------------------------------------------------------

  static const String organizations = '/api/v1/organizations';

  static String organization(String id) => '/api/v1/organizations/$id';

  // ---------------------------------------------------------------------------
  // Locations
  // ---------------------------------------------------------------------------

  static const String locations = '/api/v1/locations';

  static String location(String id) => '/api/v1/locations/$id';

  // ---------------------------------------------------------------------------
  // Sync
  // ---------------------------------------------------------------------------

  static const String syncStatus = '/api/v1/sync/status';
  static const String syncPeers = '/api/v1/sync/peers';
  static const String syncTrigger = '/api/v1/sync/trigger';
  static const String syncHistory = '/api/v1/sync/history';
  static const String syncBundleExport = '/api/v1/sync/bundle/export';
  static const String syncBundleImport = '/api/v1/sync/bundle/import';

  // ---------------------------------------------------------------------------
  // Conflicts
  // ---------------------------------------------------------------------------

  static const String conflicts = '/api/v1/conflicts';

  static String conflict(String id) => '/api/v1/conflicts/$id';
  static String conflictResolve(String id) => '/api/v1/conflicts/$id/resolve';
  static String conflictDefer(String id) => '/api/v1/conflicts/$id/defer';

  // ---------------------------------------------------------------------------
  // Alerts
  // ---------------------------------------------------------------------------

  static const String alerts = '/api/v1/alerts';
  static const String alertsSummary = '/api/v1/alerts/summary';

  static String alert(String id) => '/api/v1/alerts/$id';
  static String alertAcknowledge(String id) =>
      '/api/v1/alerts/$id/acknowledge';
  static String alertDismiss(String id) => '/api/v1/alerts/$id/dismiss';

  // ---------------------------------------------------------------------------
  // Formulary
  // ---------------------------------------------------------------------------

  static const String formularyMedications = '/api/v1/formulary/medications';

  static String formularyMedicationsByCategory(String category) =>
      '/api/v1/formulary/medications/category/$category';

  static String formularyMedication(String code) =>
      '/api/v1/formulary/medications/$code';

  static const String formularyCheckInteractions =
      '/api/v1/formulary/check-interactions';
  static const String formularyCheckAllergies =
      '/api/v1/formulary/check-allergies';
  static const String formularyDosingValidate =
      '/api/v1/formulary/dosing/validate';
  static const String formularyDosingOptions =
      '/api/v1/formulary/dosing/options';
  static const String formularyDosingSchedule =
      '/api/v1/formulary/dosing/schedule';

  static String formularyStock(String siteId, String medicationCode) =>
      '/api/v1/formulary/stock/$siteId/$medicationCode';

  static String formularyStockPrediction(
          String siteId, String medicationCode) =>
      '/api/v1/formulary/stock/$siteId/$medicationCode/prediction';

  static const String formularyDeliveries = '/api/v1/formulary/deliveries';

  static String formularyRedistribution(String medicationCode) =>
      '/api/v1/formulary/redistribution/$medicationCode';

  static const String formularyInfo = '/api/v1/formulary/info';

  // ---------------------------------------------------------------------------
  // Anchor
  // ---------------------------------------------------------------------------

  static const String anchorStatus = '/api/v1/anchor/status';
  static const String anchorVerify = '/api/v1/anchor/verify';
  static const String anchorHistory = '/api/v1/anchor/history';
  static const String anchorTrigger = '/api/v1/anchor/trigger';
  static const String anchorDidNode = '/api/v1/anchor/did/node';

  static String anchorDidDevice(String deviceId) =>
      '/api/v1/anchor/did/device/$deviceId';

  static const String anchorDidResolve = '/api/v1/anchor/did/resolve';
  static const String anchorCredentialsIssue =
      '/api/v1/anchor/credentials/issue';
  static const String anchorCredentialsVerify =
      '/api/v1/anchor/credentials/verify';
  static const String anchorCredentials = '/api/v1/anchor/credentials';
  static const String anchorBackends = '/api/v1/anchor/backends';

  static String anchorBackend(String name) =>
      '/api/v1/anchor/backends/$name';

  static const String anchorQueue = '/api/v1/anchor/queue';

  // ---------------------------------------------------------------------------
  // Supply
  // ---------------------------------------------------------------------------

  static const String supplyInventory = '/api/v1/supply/inventory';

  static String supplyInventoryItem(String itemCode) =>
      '/api/v1/supply/inventory/$itemCode';

  static const String supplyDeliveries = '/api/v1/supply/deliveries';
  static const String supplyPredictions = '/api/v1/supply/predictions';
  static const String supplyRedistribution = '/api/v1/supply/redistribution';

  // ---------------------------------------------------------------------------
  // SMART on FHIR
  // ---------------------------------------------------------------------------

  static const String smartClients = '/api/v1/smart/clients';

  static String smartClient(String id) => '/api/v1/smart/clients/$id';
}

File: open-nucleus-app/src/lib/api-paths.ts
/**
 * All REST API endpoint paths for the Open Nucleus Go backend.
 * Mirrors the chi router defined in internal/router/router.go.
 */
export const API = {
  // ---------- public ----------
  health: '/health',
  fhirMetadata: '/fhir/metadata',
  smartConfiguration: '/.well-known/smart-configuration',

  // ---------- auth ----------
  auth: {
    login: '/api/v1/auth/login',
    refresh: '/api/v1/auth/refresh',
    logout: '/api/v1/auth/logout',
    whoami: '/api/v1/auth/whoami',
  },

  // ---------- SMART OAuth2 ----------
  smart: {
    authorize: '/auth/smart/authorize',
    token: '/auth/smart/token',
    revoke: '/auth/smart/revoke',
    introspect: '/auth/smart/introspect',
    register: '/auth/smart/register',
    launch: '/auth/smart/launch',
  },

  // ---------- patients ----------
  patients: {
    list: '/api/v1/patients',
    create: '/api/v1/patients',
    search: '/api/v1/patients/search',
    match: '/api/v1/patients/match',
    get: (id: string) => `/api/v1/patients/${id}` as const,
    update: (id: string) => `/api/v1/patients/${id}` as const,
    delete: (id: string) => `/api/v1/patients/${id}` as const,
    history: (id: string) => `/api/v1/patients/${id}/history` as const,
    timeline: (id: string) => `/api/v1/patients/${id}/timeline` as const,
    erase: (id: string) => `/api/v1/patients/${id}/erase` as const,

    // encounters
    encounters: {
      list: (patientId: string) =>
        `/api/v1/patients/${patientId}/encounters` as const,
      create: (patientId: string) =>
        `/api/v1/patients/${patientId}/encounters` as const,
      get: (patientId: string, encounterId: string) =>
        `/api/v1/patients/${patientId}/encounters/${encounterId}` as const,
      update: (patientId: string, encounterId: string) =>
        `/api/v1/patients/${patientId}/encounters/${encounterId}` as const,
    },

    // observations
    observations: {
      list: (patientId: string) =>
        `/api/v1/patients/${patientId}/observations` as const,
      create: (patientId: string) =>
        `/api/v1/patients/${patientId}/observations` as const,
      get: (patientId: string, observationId: string) =>
        `/api/v1/patients/${patientId}/observations/${observationId}` as const,
    },

    // conditions
    conditions: {
      list: (patientId: string) =>
        `/api/v1/patients/${patientId}/conditions` as const,
      create: (patientId: string) =>
        `/api/v1/patients/${patientId}/conditions` as const,
      update: (patientId: string, conditionId: string) =>
        `/api/v1/patients/${patientId}/conditions/${conditionId}` as const,
    },

    // medication requests
    medicationRequests: {
      list: (patientId: string) =>
        `/api/v1/patients/${patientId}/medication-requests` as const,
      create: (patientId: string) =>
        `/api/v1/patients/${patientId}/medication-requests` as const,
      update: (patientId: string, medicationRequestId: string) =>
        `/api/v1/patients/${patientId}/medication-requests/${medicationRequestId}` as const,
    },

    // allergy intolerances
    allergyIntolerances: {
      list: (patientId: string) =>
        `/api/v1/patients/${patientId}/allergy-intolerances` as const,
      create: (patientId: string) =>
        `/api/v1/patients/${patientId}/allergy-intolerances` as const,
      update: (patientId: string, allergyId: string) =>
        `/api/v1/patients/${patientId}/allergy-intolerances/${allergyId}` as const,
    },

    // immunizations
    immunizations: {
      list: (patientId: string) =>
        `/api/v1/patients/${patientId}/immunizations` as const,
      create: (patientId: string) =>
        `/api/v1/patients/${patientId}/immunizations` as const,
      get: (patientId: string, immunizationId: string) =>
        `/api/v1/patients/${patientId}/immunizations/${immunizationId}` as const,
    },

    // procedures
    procedures: {
      list: (patientId: string) =>
        `/api/v1/patients/${patientId}/procedures` as const,
      create: (patientId: string) =>
        `/api/v1/patients/${patientId}/procedures` as const,
      get: (patientId: string, procedureId: string) =>
        `/api/v1/patients/${patientId}/procedures/${procedureId}` as const,
    },

    // consents (patient-scoped)
    consents: {
      list: (patientId: string) =>
        `/api/v1/patients/${patientId}/consents` as const,
      grant: (patientId: string) =>
        `/api/v1/patients/${patientId}/consents` as const,
    },
  },

  // ---------- consent management (top-level) ----------
  consents: {
    revoke: (consentId: string) =>
      `/api/v1/consents/${consentId}` as const,
    issueVC: (consentId: string) =>
      `/api/v1/consents/${consentId}/vc` as const,
  },

  // ---------- practitioners ----------
  practitioners: {
    list: '/api/v1/practitioners',
    create: '/api/v1/practitioners',
    get: (id: string) => `/api/v1/practitioners/${id}` as const,
    update: (id: string) => `/api/v1/practitioners/${id}` as const,
  },

  // ---------- organizations ----------
  organizations: {
    list: '/api/v1/organizations',
    create: '/api/v1/organizations',
    get: (id: string) => `/api/v1/organizations/${id}` as const,
  },

  // ---------- locations ----------
  locations: {
    list: '/api/v1/locations',
    create: '/api/v1/locations',
    get: (id: string) => `/api/v1/locations/${id}` as const,
  },

  // ---------- sync ----------
  sync: {
    status: '/api/v1/sync/status',
    peers: '/api/v1/sync/peers',
    trigger: '/api/v1/sync/trigger',
    history: '/api/v1/sync/history',
    exportBundle: '/api/v1/sync/bundle/export',
    importBundle: '/api/v1/sync/bundle/import',
  },

  // ---------- conflicts ----------
  conflicts: {
    list: '/api/v1/conflicts',
    get: (id: string) => `/api/v1/conflicts/${id}` as const,
    resolve: (id: string) => `/api/v1/conflicts/${id}/resolve` as const,
    defer: (id: string) => `/api/v1/conflicts/${id}/defer` as const,
  },

  // ---------- alerts ----------
  alerts: {
    list: '/api/v1/alerts',
    summary: '/api/v1/alerts/summary',
    get: (id: string) => `/api/v1/alerts/${id}` as const,
    acknowledge: (id: string) => `/api/v1/alerts/${id}/acknowledge` as const,
    dismiss: (id: string) => `/api/v1/alerts/${id}/dismiss` as const,
  },

  // ---------- formulary ----------
  formulary: {
    medications: '/api/v1/formulary/medications',
    medicationsByCategory: (category: string) =>
      `/api/v1/formulary/medications/category/${category}` as const,
    medication: (code: string) =>
      `/api/v1/formulary/medications/${code}` as const,
    checkInteractions: '/api/v1/formulary/check-interactions',
    checkAllergies: '/api/v1/formulary/check-allergies',
    dosingValidate: '/api/v1/formulary/dosing/validate',
    dosingOptions: '/api/v1/formulary/dosing/options',
    dosingSchedule: '/api/v1/formulary/dosing/schedule',
    stockLevel: (siteId: string, medicationCode: string) =>
      `/api/v1/formulary/stock/${siteId}/${medicationCode}` as const,
    updateStock: (siteId: string, medicationCode: string) =>
      `/api/v1/formulary/stock/${siteId}/${medicationCode}` as const,
    stockPrediction: (siteId: string, medicationCode: string) =>
      `/api/v1/formulary/stock/${siteId}/${medicationCode}/prediction` as const,
    deliveries: '/api/v1/formulary/deliveries',
    redistribution: (medicationCode: string) =>
      `/api/v1/formulary/redistribution/${medicationCode}` as const,
    info: '/api/v1/formulary/info',
  },

  // ---------- anchor ----------
  anchor: {
    status: '/api/v1/anchor/status',
    verify: '/api/v1/anchor/verify',
    history: '/api/v1/anchor/history',
    trigger: '/api/v1/anchor/trigger',
    didNode: '/api/v1/anchor/did/node',
    didDevice: (deviceId: string) =>
      `/api/v1/anchor/did/device/${deviceId}` as const,
    didResolve: '/api/v1/anchor/did/resolve',
    credentialsIssue: '/api/v1/anchor/credentials/issue',
    credentialsVerify: '/api/v1/anchor/credentials/verify',
    credentialsList: '/api/v1/anchor/credentials',
    backends: '/api/v1/anchor/backends',
    backend: (name: string) => `/api/v1/anchor/backends/${name}` as const,
    queue: '/api/v1/anchor/queue',
  },

  // ---------- supply ----------
  supply: {
    inventory: '/api/v1/supply/inventory',
    inventoryItem: (itemCode: string) =>
      `/api/v1/supply/inventory/${itemCode}` as const,
    deliveries: '/api/v1/supply/deliveries',
    predictions: '/api/v1/supply/predictions',
    redistribution: '/api/v1/supply/redistribution',
  },

  // ---------- SMART client management ----------
  smartClients: {
    list: '/api/v1/smart/clients',
    get: (id: string) => `/api/v1/smart/clients/${id}` as const,
    update: (id: string) => `/api/v1/smart/clients/${id}` as const,
    delete: (id: string) => `/api/v1/smart/clients/${id}` as const,
  },

  // ---------- FHIR R4 REST ----------
  fhir: {
    search: (resourceType: string) => `/fhir/${resourceType}` as const,
    read: (resourceType: string, id: string) =>
      `/fhir/${resourceType}/${id}` as const,
    create: (resourceType: string) => `/fhir/${resourceType}` as const,
    update: (resourceType: string, id: string) =>
      `/fhir/${resourceType}/${id}` as const,
    delete: (resourceType: string, id: string) =>
      `/fhir/${resourceType}/${id}` as const,
  },

  // ---------- WebSocket (stub) ----------
  ws: '/api/v1/ws',
} as const;
